haystack/nodes/retriever/sparse.py

```diff
@@ -606,41 +609,41 @@ def retrieve_batch(
                             Otherwise raw similarity scores (e.g. cosine or dot_product) will be used.
         :param document_store: the docstore to use for retrieval. If `None`, the one given in the `__init__` is used instead.
         """
+        if filters:
+            raise NotImplementedError("Filters are not implemented in TfidfRetriever.")
+        if scale_score:
+            raise NotImplementedError("Scaling score to the unit interval is not supported in TfidfRetriever.")
+
+        document_store = document_store or self.document_store
         if document_store is None:
-            document_store = self.document_store
-            if document_store is None:
-                raise ValueError(
-                    "This Retriever was not initialized with a Document Store. Provide one to the retrieve() method."
-                )
-        else:
-            self.fit(document_store=document_store)
+            raise ValueError(
+                "This Retriever was not initialized with a Document Store. Provide one to the retrieve() method."
+            )
+        index = index or document_store.index
+        assert isinstance(index, str)  # Necessary for mypy
 
         if self.auto_fit:
-            if document_store.get_document_count(headers=headers) != self.document_count:
-                # run fit() to update self.df, self.tfidf_matrix and self.document_count
+            if (
+                index not in self.document_counts
+                or document_store.get_document_count(headers=headers, index=index) != self.document_counts[index]
+            ):
+                # run fit() to update self.dataframes, self.tfidf_matrices and self.document_counts
                 logger.warning(
                     "Indexed documents have been updated and fit() method needs to be run before retrieval. Running it now."
                 )
-                self.fit(document_store=document_store)
-        if self.df is None:
+                self.fit(document_store=document_store, index=index)
+        if self.dataframes[index] is None:
             raise DocumentStoreError(
-                "Retrieval requires dataframe df and tf-idf matrix but fit() did not calculate them probably due to an empty document store."
+                "Retrieval requires dataframe and tf-idf matrix but fit() did not calculate them probably due to an empty document store."
```

Comment:
```suggestion
                "Retrieval requires dataframe and tf-idf matrix but fit() did not calculate them probably because of an empty document store."
```